### DISCUSSION TITLE
[Discussion] Adding features to related notes

### DISCUSSION THREAD

**Discussion by chuyouchia** [reorproject/reor#45] [Ideas]
# Some thoughts on related notes

The current way related notes is implemented actually is pretty useful, but imo could be better structured to allow users to do things on top of related notes. Some things I hope to see:

*   If we are to leverage the power of semantic linking, creating links between chunks of information (similar to Obsidian/Notion), rather than the whole note would be amazing. In practice, this might be window focus (or highlighting of specific notes), then on focus, using that highlighted chunk to automatically suggesting linking to other relevant notes. Or it could be a copilot-esque suggestion that appears as the user types. If users agree that they are related, it should lead to linking of the chunk of current note to the suggested note.&#x20;
*   Once that bidirectional linking is established, ideally it is added as an inline preview to the note. At that point we should remove it from the related notes list, as it is no longer a just suggestion, but an accepted suggestion. In the related notes, I think I only want to see suggested notes, rather than ones that I have already accepted, to reduce mental load or the need for scrolling. Kinda like checking off todos off the related notes list
*   On the basis of that linking, it would also be useful to organize the notes into a mind-map based off the links, rather than linear notes like what the current related notes tab offer. This would help with review of material. Think a medical school student or CS student studying for their exams needing to make connections between old source material and new ones. On that note, auto generated flashcards would probably make this useful for study purposes too.
*   Another use case instead of linking notes is to allow users to bring related notes into existing notes and combine them. 2 discrete notes might be the result of scattered brainstorming on related topics. If the related notes tab has a copy feature, that would be very useful imo. Even better if it would preserve the markdown syntax when copying notes from the related notes tab.

Let me know what do you think and if these can be added to the roadmap, I am hoping for this local LLM note taker to advance and want to contribute in any way possible, mostly because I would love to use something like this. Currently a full stack dev with about 2+ years of experience with React/Nestjs/Typescript, so probably able to help out with the feature development and UI aspects of the app. At my current day job, I am a backend dev so I can help with backend tasks too.

**Comment by samlhuillier** [comment]
Thank you for raising this. I really appreciate how much thought you've put into this.

I agree that the sidebar is valuable but not quite there. Almost a proof of concept with lots of potential to be super useful.

On each of your points:

- Yes the next step will be linking between chunks. Chunking is already implemented in the backend, it's just a case of deciding how to display the semantic links corresponding to each chunk in the editor. Some ideas I have on this are highlighting/selecting blocks & multiple blocks reveals on the sidebar only the semantic links corresponding to what is selected. Inline related notes could also be an idea here. In the latter part of this first point, do you mean that a copilot suggestion would then becomes its own standalone note? (Related to this too is the idea of a notion-like writing assistant which is also something on the roadmap)

- I like this idea of checking off todos from related thoughts as if to be like ensuring you've thought of everything based on your previous thoughts. And this'd be flexible - maybe we could even have a toggle button in the sidebar to toggle between showing bidirectional links or semantic links.

- Yes mind maps is a nice feature of course and definitely coming. Given everything is embedded as a multi dimensional semantic vector there are opportunities here to combine a T-SNE graph with an Obsidian-like graphs of links. I feel like this could be quite powerful.

- Yes combining notes with like drag and drop/copy would be a super useful way to consolidate supporting ideas from different places.

Another point on bidirectional/manual linking is that I think a very powerful workflow will be the same square bracket workflow like Obsidian but suggestions being semantic rather than based on filenames.

I'd love to chat more about how you could contribute more to the project! Please feel free to reach out via email: samlhuillier1@gmail.com

Or if you prefer to [schedule a call](https://cal.com/lhuillier).

I'm also going to be putting together an initial roadmap which it would be great to get your input on.

**Reply by chuyouchia** [reply]
Let me try to explain a little.
1.  **_Do you mean that a copilot suggestion would then becomes its own standalone note?_** I meant that it could be a copilot-esque suggestion that pops up to ask the user if they want to link the notes. Not necessarily to generate a new standalone note.
2. Yeah the toggle to decide whether to include suggested notes or not sounds like a good idea too. Maybe this is something we trial and find out from user testing. Another way we can make the related notes very useful IMO is the concept of a citation link in research papers, where you can click and focus on a linked note (or citation number) and it highlights the linked note in the sidebar.
3. Yeah love the combination of suggested T-SNE graph with Obsidian link graph of links into a mindmap! Would probably need to find a way to visually distinguish between the two though. 
4. Awesome

Sent you an email, title is `Reor Collaboration`!

**Comment by RazyRo** [comment]
> 3\. Obsidian link graph
may i ask why t-SNE instead of UMAP?

**Reply by samlhuillier** [reply]
No strong preference actually. We'd have to investigate this. Right now though, we haven't been thinking about visualising the embeddings into a graph. Is that something you'd be interested in? Is there a usecase for which you think this'd be relevant?

**Reply by RazyRo** [reply]
i am not exactly sure the reason why t-SNE or UMAP.
But I think using embeddings into a graph might be useful for link prediction and node classification. By analysing the relationships between nodes in the embedding space, UMAP/t-SNE can help predict potential links that may exist between nodes that haven't been explicitly connected in the original graph. The embedding captures the relevant information about a node's position and relationships within the graph, which can be beneficial for classification algorithms.
Example: You wouldn't use the 2D map to navigate the city, but you would use the intersection summaries (embeddings) to build a navigation system that predicts traffic flow (link prediction) or classifies a street as residential or commercial (node classification).

**Reply by BradKML** [reply]
Wait aren't UMAP reversible compared to t-SNE? And also how some of the embeddings like Trimap or PaCMAP or LargeVis can be better or faster sometimes in certain context?
